### PR TITLE
Update all projects to TypeScript 2.0 - remove typings - etc

### PR DESCRIPTION
**Pull request by MarkPieszak**
All projects up & running now with TS2.0.

### CODE REVIEW

File: chokidar/package.json [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      {
+        "raw": "chokidar@^1.0.0",
+        "scope": null,
+        "escapedName": "chokidar",
+        "name": "chokidar",
+        "rawSpec": "^1.0.0",
+        "spec": ">=1.0.0 <2.0.0",
+        "type": "range"
+      },
+      "C:\\Users\\mark.pieszak\\Documents\\GitHub\\JavaScriptServices\\templates\\KnockoutSpa\\node_modules\\watchpack"
```
Comment (tinchou) [inline chokidar/package.json]: This is probably not right (your path). But what is chokidar doing here anyway?

Comment (MarkPieszak) [inline chokidar/package.json]: I'm not sure removing it now... just realized that too

**Comment by SteveSandersonMS** [comment]
Thanks for the update @MarkPieszak!

There seems to be some unrelated stuff in the PR that was perhaps added by mistake. For example, the `chokidar` source code for some reason, and renaming `boot-client.ts` to `main-client.ts` etc. 

Could you remove the unrelated changes from this PR so that it's just the updates for TS 2.0?

File: templates/KnockoutSpa/ClientApp/boot.ts [hunk 2]
```diff
@@ -1,8 +1,8 @@
 import './css/site.css';
 import * as ko from 'knockout';
-import { createHistory } from 'history';
 import './webpack-component-loader';
 import AppRootComponent from './components/app-root/app-root';
+var createHistory = require('history').createBrowserHistory
```
Comment (SteveSandersonMS) [inline templates/KnockoutSpa/ClientApp/boot.ts]: Is this change deliberate? Could you clarify why it would be needed (hopefully it isn't), and if it is, change the `var` to `const` and add the semicolon at the end of the line?

File: templates/KnockoutSpa/tsconfig.json [hunk 3]
```diff
@@ -3,8 +3,13 @@
...
     "target": "es5",
     "sourceMap": true,
-    "skipDefaultLibCheck": true
+    "skipDefaultLibCheck": true,
+    "typeRoots": [ "node_modules/@types" ],
+    "types": ["core-js", "crossroads", "history", "knockout",
+     "requirejs", "signals", "whatwg-fetch" ]
   },
+  "compileOnSave": false,
+  "buildOnSave": false,
```
Comment (SteveSandersonMS) [inline templates/KnockoutSpa/tsconfig.json]: Are these two lines really needed?

File: templates/ReactReduxSpa/ClientApp/configureStore.ts [hunk 4]
```diff
@@ -1,27 +1,29 @@
-import { createStore, applyMiddleware, compose, combineReducers } from 'redux';
-import * as thunkModule from 'redux-thunk';
-import { routerReducer } from 'react-router-redux';
-import * as Store from './store';
-import { typedToPlain } from 'redux-typed';
+import * as Redux from "redux";
```
Comment (SteveSandersonMS) [inline templates/ReactReduxSpa/ClientApp/configureStore.ts]: For consistency, can we go back to single-quotes for all lines in this file?

File: templates/ReactReduxSpa/ClientApp/configureStore.ts [hunk 5]
```diff
@@ -1,27 +1,29 @@
...
+import { createStore, applyMiddleware, compose, combineReducers, GenericStoreEnhancer } from "redux";
+import thunk from "redux-thunk";
+import { routerReducer } from "react-router-redux";
+import * as Store from "./store";
+import { typedToPlain } from "redux-typed";
+
```
Comment (SteveSandersonMS) [inline templates/ReactReduxSpa/ClientApp/configureStore.ts]: There's already a blank line below this - don't need another :)

File: templates/ReactReduxSpa/ClientApp/configureStore.ts (hunk 4 above)
Comment (MarkPieszak) [inline templates/ReactReduxSpa/ClientApp/configureStore.ts]: Sure!

File: templates/ReactReduxSpa/ClientApp/configureStore.ts (hunk 5 above)
Comment (MarkPieszak) [inline templates/ReactReduxSpa/ClientApp/configureStore.ts]: Haha sounds good, be right back (grabbing lunch) and I'll fix the above comments :+1:

File: templates/ReactReduxSpa/tsconfig.json [hunk 6]
```diff
@@ -5,7 +5,12 @@
     "jsx": "preserve",
     "experimentalDecorators": true,
     "sourceMap": true,
-    "skipDefaultLibCheck": true
+    "skipDefaultLibCheck": true,
+    "lib": ["es6", "dom"],
+    "typeRoots": [ "node_modules/@types" ],
```
Comment (SteveSandersonMS) [inline templates/ReactReduxSpa/tsconfig.json]: Is this needed?

File: templates/KnockoutSpa/ClientApp/components/app-root/app-root.ts [hunk 7]
```diff
@@ -40,4 +40,4 @@ class AppRootViewModel {
     }
 }
 
-export default { viewModel: AppRootViewModel, template: require('./app-root.html') };
+export default { viewModel: AppRootViewModel, componentName: 'Home', template: require('./app-root.html') };
```
Comment (MarkPieszak) [inline templates/KnockoutSpa/ClientApp/components/app-root/app-root.ts]: This might of been accidental, Knockout is giving me a minor :
Any idea what that might be? @SteveSandersonMS 

```vendor.js?v=P985CJI4_0EhN8wTjSzDXZnMQt78MgU91Uz14XJKMk8:26089 
Uncaught Error: Unable to process binding 
"component: function (){return { name:route().page,params:route} }"
Message: No component name specified```

File: chokidar/package.json (hunk 1 above)
Comment (MarkPieszak) [inline chokidar/package.json]: Removed it now!

File: templates/KnockoutSpa/ClientApp/boot.ts (hunk 2 above)
Comment (MarkPieszak) [inline templates/KnockoutSpa/ClientApp/boot.ts]: Be back shortly and i'll talk to you about this one.

File: templates/KnockoutSpa/ClientApp/components/app-root/app-root.ts (hunk 7 above)
Comment (SteveSandersonMS) [inline templates/KnockoutSpa/ClientApp/components/app-root/app-root.ts]: Sounds like the URL isn't matching any route, so `route()` is an empty object, so `route().page` is undefined, so you're returning nothing for the `name` of the component.

File: templates/Angular2Spa/webpack.config.js [hunk 8]
```diff
@@ -41,7 +41,7 @@ var clientBundleConfig = merge(sharedConfig, {
 
 // Configuration for server-side (prerendering) bundle suitable for running in Node
 var serverBundleConfig = merge(sharedConfig, {
-    entry: { 'main-server': './ClientApp/boot-server.ts' },
+    entry: { 'main-server': './ClientApp/main-server.ts' },
```
Comment (SteveSandersonMS) [inline templates/Angular2Spa/webpack.config.js]: Let's keep renaming separate from this PR. Can you revert the names?

File: templates/ReactReduxSpa/tsconfig.json [hunk 9]
```diff
@@ -5,7 +5,12 @@
...
+    "types": [ "react", "react-dom", "react-redux", "react-router", "react-router-redux",
+     "redux", "redux-thunk", "source-map", "uglify-js", "webpack", "webpack/webpack-env",
```
Comment (stephtr) [inline templates/ReactReduxSpa/tsconfig.json]: Even though it is not yet at npm: `webpack-env` has been outsourced into an additional package ([DefinitelyTyped #11684](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/11684)).

File: templates/ReactReduxSpa/ClientApp/configureStore.ts [hunk 10]
```diff
@@ -1,27 +1,29 @@
...
 
 export default function configureStore(initialState?: Store.ApplicationState) {
     // Build middleware. These are functions that can process the actions before they reach the store.
-    const thunk = (thunkModule as any).default; // Workaround for TypeScript not importing thunk module as expected
-    const windowIfDefined = typeof window === 'undefined' ? null : window as any;
-    const devToolsExtension = windowIfDefined && windowIfDefined.devToolsExtension; // If devTools is installed, connect to it
+    const windowIfDefined = typeof window === "undefined" ? null : window as any;
+    // If devTools is installed, connect to it
+    const devToolsExtension = windowIfDefined && windowIfDefined.devToolsExtension as () => GenericStoreEnhancer;
     const createStoreWithMiddleware = compose(
```
Comment (stephtr) [inline templates/ReactReduxSpa/ClientApp/configureStore.ts]: Not relevant yet, but I wanted to note it somewhere: This line won't work with the most recent version of redux.d.ts ([redux #1936](https://github.com/reactjs/redux/pull/1936#issuecomment-250443020)) and propably have to be changed in the future to `compose<Redux.StoreEnhancerStoreCreator<Store.ApplicationState>>(`, which doesn't work with the version available now on npm.

File: templates/ReactReduxSpa/ClientApp/configureStore.ts [hunk 11]
```diff
@@ -1,27 +1,29 @@
-import { createStore, applyMiddleware, compose, combineReducers } from 'redux';
-import * as thunkModule from 'redux-thunk';
-import { routerReducer } from 'react-router-redux';
-import * as Store from './store';
-import { typedToPlain } from 'redux-typed';
+import * as Redux from "redux";
+import { createStore, applyMiddleware, compose, combineReducers, GenericStoreEnhancer } from "redux";
```
Comment (MarkPieszak) [inline templates/ReactReduxSpa/ClientApp/configureStore.ts]: Just noticed that, somehow overlooked that line too! :P

File: templates/Angular2Spa/webpack.config.js (hunk 8 above)
Comment (MarkPieszak) [inline templates/Angular2Spa/webpack.config.js]: It was breaking since webpack is looking for a file by that name, want me to change it there instead?

File: templates/KnockoutSpa/tsconfig.json (hunk 3 above)
Comment (MarkPieszak) [inline templates/KnockoutSpa/tsconfig.json]: They are just helpful depending on certain IDE's to disable automatic `tsc` compilation / etc.
Do you want me to remove them? Just thought they were helpful to ensure webpack handles everything.

File: templates/KnockoutSpa/ClientApp/components/app-root/app-root.ts [hunk 12]
```diff
@@ -4,7 +4,7 @@ import navMenu from '../nav-menu/nav-menu';
 
 // Declare the client-side routing configuration
 const routes: Route[] = [
-    { url: '',              params: { page: 'home-page' } },
+    { url: '/',              params: { page: 'home-page' } },
```
Comment (MarkPieszak) [inline templates/KnockoutSpa/ClientApp/components/app-root/app-root.ts]: This was an attempt to fix a knockout issue, I can revert this change!

File: templates/ReactReduxSpa/tsconfig.json (hunk 9 above)
Comment (stephtr) [inline templates/ReactReduxSpa/tsconfig.json]: Theoretically it should be already online. I filed an issue at [types-publisher #159](https://github.com/Microsoft/types-publisher/issues/159).

File: templates/ReactReduxSpa/tsconfig.json (hunk 6 above)
Comment (MarkPieszak) [inline templates/ReactReduxSpa/tsconfig.json]: Not specifically, it's just nice to see at a glance what Type definitions are being used.

File: templates/KnockoutSpa/ClientApp/components/app-root/app-root.ts (hunk 7 above)
Comment (MarkPieszak) [inline templates/KnockoutSpa/ClientApp/components/app-root/app-root.ts]: So when a refresh is done it doesn't load that routed content, but when the link is clicked, it works fine. Any idea what could be causing this?

File: templates/KnockoutSpa/ClientApp/boot.ts (hunk 2 above)
Comment (MarkPieszak) [inline templates/KnockoutSpa/ClientApp/boot.ts]: So it needs to be updated because when the previous `import {}` way was done, with the new History, it doesn't call the correct methods and ends up erroring that "createBrowserHistory" doesn't exist. I've tried the other ways suggested on the packages readme (https://github.com/mjackson/history) but only this one works). Updating it to `const` now.

File: templates/KnockoutSpa/tsconfig.json [hunk 13]
```diff
@@ -3,8 +3,13 @@
     "moduleResolution": "node",
     "target": "es5",
     "sourceMap": true,
-    "skipDefaultLibCheck": true
+    "skipDefaultLibCheck": true,
+    "typeRoots": [ "node_modules/@types" ],
```
Comment (MarkPieszak) [inline templates/KnockoutSpa/tsconfig.json]: Removed the `typeRoots` part: (`node_modules/@types`) is default.
The part where `types:[]` is set is needed otherwise it won't make them necessary during compilation.

File: templates/ReactReduxSpa/ClientApp/configureStore.ts (hunk 11 above)
Comment (MarkPieszak) [inline templates/ReactReduxSpa/ClientApp/configureStore.ts]: Fixed.

**Comment by MarkPieszak** [comment]
@SteveSandersonMS This should be all set now, minus the content not loading right away in the KnockoutSpa template (maybe you know what's causing it?)

Removed the typeRoots, changed createHistry to `const`
Eliminated double Enters, Double quotes on the one page, etc.

**Comment by stephtr** [comment]
Just for information: The redux-typed/StrongActions.d.ts error is still remaining.

File: templates/KnockoutSpa/tsconfig.json (hunk 3 above)
Comment (stephtr) [inline templates/KnockoutSpa/tsconfig.json]: Why are they only in one template and not all?

File: templates/KnockoutSpa/ClientApp/boot.ts (hunk 2 above)
Comment (stephtr) [inline templates/KnockoutSpa/ClientApp/boot.ts]: It seems to be connected to tsconfig.json "target": "es5". Switching to es6 results in an error with crossroads.

**Comment by MarkPieszak** [comment]
Yes, strangely the error comes from within node_modules itself. Made an issue, but seems no one has responded to it, also doesn't seem to prevent anything from working luckily!

File: templates/ReactReduxSpa/tsconfig.json (hunk 9 above)
Comment (stephtr) [inline templates/ReactReduxSpa/tsconfig.json]: @types/webpack-env is online now.

File: templates/ReactReduxSpa/tsconfig.json [hunk 14]
```diff
@@ -5,7 +5,11 @@
...
     "experimentalDecorators": true,
     "sourceMap": true,
-    "skipDefaultLibCheck": true
+    "skipDefaultLibCheck": true,
+    "lib": ["es6", "dom"],
+    "types": [ "react", "react-dom", "react-redux", "react-router", "react-router-redux",
+     "redux", "redux-thunk", "source-map", "uglify-js", "webpack", "webpack/webpack-env",
+      "whatwg-fetch" ]
   },
   "exclude": [
```
Comment (antmdvs) [inline templates/ReactReduxSpa/tsconfig.json]: @SteveSandersonMS Should `wwwroot` be excluded as well? If so, can we sneak that in this PR?

File: templates/ReactReduxSpa/ClientApp/configureStore.ts (hunk 10 above)
Comment (stephtr) [inline templates/ReactReduxSpa/ClientApp/configureStore.ts]: Just another note on these lines, they could be replaced by (like in the [redux documentation](http://redux.js.org/docs/api/compose.html)):
```
const store = createStore(
	buildRootReducer(Store.reducers),
	initialState,
	compose(
		applyMiddleware(thunk, typedToPlain),
		devToolsExtension ? devToolsExtension() : f => f
	)
);
```
Which also eliminates one typecast.

File: templates/ReactReduxSpa/tsconfig.json (hunk 9 above)
Comment (MarkPieszak) [inline templates/ReactReduxSpa/tsconfig.json]: great, i'll replace `webpack/webpack-env` in here with that `@types/webpack-env`. Thanks @stephtr !

File: templates/ReactReduxSpa/ClientApp/configureStore.ts (hunk 10 above)
Comment (stephtr) [inline templates/ReactReduxSpa/ClientApp/configureStore.ts]: `compose(...) as StoreEnhancer<Store.ApplicationState>` would be an option which is compatible with redux 3.6.0` **and** upcoming versions.

**Comment by stephtr** [comment]
As far as I noticed there is a `declare var module: any;` in `ReactReduxSpa/ClientApp/routes.tsx`, which isn't needed anymore.

File: templates/ReactReduxSpa/tsconfig.json (hunk 14 above)
Comment (antmdvs) [inline templates/ReactReduxSpa/tsconfig.json]: I defer to @SteveSandersonMS, as I'm not even sure if it's necessary. Just something I noticed.

**Comment by MarkPieszak** [comment]
@SteveSandersonMS Everything has been updated to the latest from master.
The only issue remaining (as far as I've seen) is KnockoutSpa not painting that initial Route if you could take a look?

I didn't want to add angular2 recent upgrades (2.1 & universal 2.1 patches) until we were finished with this one yet. Let me know what you think!

**Comment by emzero** [comment]
The `StrongActions.d.ts` error is easily fixed by setting the `dispatch` argument type to `Dispatch<TState>`, but since Redux now includes its own typings, wouldn't be better to leverage them so we stay as close to "official" as possible? 
Also, that `redux-typed` package seems quite unpopular.

**Comment by MarkPieszak** [comment]
Great idea, totally agree, rather it be as official as possible.
@radu-porumb em zeros comment above looks like what we should be doing.

**Comment by radu-porumb** [comment]
@MarkPieszak That's exactly what I did to move things along. But my project is on a CI pipeline which means as soon as it gets to source control the build server pulls it, builds it and fails spectacularly. I could always just throw out `redux-typed` and copy-paste the code with the `<TState>` fix into a local file to keep everything strongly typed but I'd like to know if we're likely to get a package update soon before going that route.

File: templates/KnockoutSpa/tsconfig.json (hunk 3 above)
Comment (SteveSandersonMS) [inline templates/KnockoutSpa/tsconfig.json]: Have removed them

**Comment by SteveSandersonMS** [comment]
@stephtr:
> As far as I noticed there is a declare var module: any; in ReactReduxSpa/ClientApp/routes.tsx, which isn't needed anymore.

Well spotted! Removed.

> Just for information: The redux-typed/StrongActions.d.ts error is still remaining.

Thanks - fixed (in newly-published `redux-typed` version 1.0.1).

@MarkPieszak:
> The only issue remaining (as far as I've seen) is KnockoutSpa not painting that initial Route if you could take a look?

The issue was that the `history` module had a breaking change between versions 2 and 4 whereby it no longer fires a navigation event on page load. I've amended the code to populate the initial history state. Unfortunately `@types/history` is massively out-of-date (it still describes the API for `history` version 2) so there's another cast-to-`any` to make this compile.

> I didn't want to add angular2 recent upgrades (2.1 & universal 2.1 patches) until we were finished with this one yet. Let me know what you think!

Makes sense. I'm merging this PR now so let's look at 2.1 soon :) Thanks.

@emzero:
> Redux now includes its own typings, wouldn't be better to leverage them so we stay as close to "official" as possible? 

Mostly I agree with you but it's a wider discussion, so let's move this part of the discussion to a separate issue.

File: templates/ReactReduxSpa/tsconfig.json (hunk 14 above)
Comment (SteveSandersonMS) [inline templates/ReactReduxSpa/tsconfig.json]: @antmdvs I'm not sure why `wwwroot` should be excluded here, but if you think there's something wrong, could you please file a separate issue specifying what problem occurs and what should be done to fix it? Thanks!

**Comment by SteveSandersonMS** [comment]
Merged! Thanks @MarkPieszak for your work on this, and for accounting for many changes while waiting for it to be merged.

**Comment by MarkPieszak** [comment]
Not a problem, glad to be a part of it all :)
Let me know if there's any other random issues you'd like me to look into? @SteveSandersonMS

**Comment by SteveSandersonMS** [comment]
> Let me know if there's any other random issues you'd like me to look into

@MarkPieszak The biggest thing for the Angular2Spa template would be if there was some way to get AoT compilation and tree-shaking. But it would have to be pretty much transparent to the developer and not make the webpack config a lot more complex. I don't know if that's possible yet, because last time I checked, `ngc` didn't have any supported way to work in a Webpack environment.

**Comment by tinchou** [comment]
It's transparent with @ngtools/webpack. But I have to admit I don't see a
great size improvement with tree shaking. Maybe I'm missing something.

I also mentioned I don't know if it'll work with Universal because I
removed it from my project.

I can share my webpack config later if you want!

**Comment by MarkPieszak** [comment]
Tinchou is right, ngtools/webpack is avaialable, but there's still a lot of things in the works to make everything more stable.

On our end, on the Universal side, we still have a little bit of ground to cover, there has been some work done to get some basic AoT going, but I think it might be a good idea to let things mature a little more before moving it over here as well? (Angular Core itself needs to mature AoT itself in a few areas)
The performance/speed improvements (so far) are incredible, so it would be amazing to get it in here :+1: 

But that sounds good to me! I'll start playing around with some things, at least open up a PR sometime so we can begin looking at webpack2's tree-shaking & AoT combination for the template.

**Comment by tinchou** [comment]
Here's my example configuration https://gist.github.com/tinchou/08d1be519b525899969180bdf7edb67f
